connector/src/main/java/com/google/cloud/spark/bigquery/ArrowBinaryIterator.java

```diff
@@ -0,0 +1,121 @@
+
+    private static long maxAllocation = Long.MAX_VALUE;
+    ArrowReaderIterator arrowReaderIterator;
+    Iterator<InternalRow> currentIterator;
+    List<String> columnsInOrder;
+
+    public ArrowBinaryIterator(List<String> columnsInOrder, ByteString schema, ByteString rowsInBytes) {
+        BufferAllocator allocator = (new RootAllocator(maxAllocation)).newChildAllocator("ArrowBinaryIterator",
+                0, maxAllocation);
+
+        SequenceInputStream bytesWithSchemaStream = new SequenceInputStream(
+                new ByteArrayInputStream(schema.toByteArray()),
+                new ByteArrayInputStream(rowsInBytes.toByteArray()));
+
+        ArrowStreamReader arrowStreamReader = new ArrowStreamReader(bytesWithSchemaStream, allocator);
+        arrowReaderIterator = new ArrowReaderIterator(arrowStreamReader);
+        this.columnsInOrder = columnsInOrder;
+        initializeCurrentIterator();
+    }
+
+    @Override
+    public boolean hasNext() {
+        initializeCurrentIterator();
+        return currentIterator.hasNext();
+    }
+
+    @Override
+    public InternalRow next() {
+        return currentIterator.next();
+    }
+
+    private void initializeCurrentIterator() {
+        while (currentIterator == null || !currentIterator.hasNext()) {
+            if (arrowReaderIterator.hasNext()) {
+                currentIterator = toArrowRows(arrowReaderIterator.next(), columnsInOrder);
+            } else {
+                break;
+            }
+            arrowReaderIterator.next();
+        }
+    }
+
+    private Iterator<InternalRow> toArrowRows(VectorSchemaRoot root, List<String> namesInOrder) {
+        List<FieldVector> vectors = namesInOrder.stream().map(name -> root.getVector(name)).collect(Collectors.toList());
+        ColumnVector[] columns = vectors.stream().map(vector -> new ArrowSchemaConverter(vector))
+                .collect(Collectors.toList()).toArray(new ColumnVector[0]);
+
+        ColumnarBatch batch = new ColumnarBatch(columns);
+        batch.setNumRows(root.getRowCount());
+        return batch.rowIterator();
+    }
+}
+
+class ArrowReaderIterator implements Iterator<VectorSchemaRoot> {
+
+    boolean closed = false;
+    VectorSchemaRoot current = null;
+    ArrowReader reader = null;
+    private static final Logger log = LoggerFactory.getLogger(AvroBinaryIterator.class);
+
+    public ArrowReaderIterator(ArrowReader reader) {
+        this.reader = reader;
+    }
+
+    @Override
+    public boolean hasNext() {
+        if (current != null) {
+            return true;
+        }
+
+        try {
+            boolean res = reader.loadNextBatch();
+            if (res) {
+                current = reader.getVectorSchemaRoot();
+            } else {
+                ensureClosed();
+            }
+            return res;
+        } catch (Exception e) {
+            log.error("Exception occurred in next() of ArrowBinaryIterator", e);
+        }
```

Comment:
@davidrabinowitz Since we cannot throw exception from java iterators, how should we handle the exception

Reply:
Which exception is it? If it is a RuntimeException subclass, you can let the client code to handle it. If it is not a RuntimeException, then convert it to one by wrapping with our own RuntimeException